### PR TITLE
Document Signature APIs in the System.​Reflection.​Metadata Namespace.

### PR DESCRIPTION
**Pull request by tompratt-AQ**
# Title
Add Undocumented APIs: System.​Reflection.​Metadata Namespace - Signature APIs

## Summary
Added missing documentation for the group of Signature APIs in the System.​Reflection.​Metadata Namespace. This group contains the following topics:

- ConstantTypeCode.xml
- CustomAttributeNamedArgument`1.xml
- CustomAttributeNamedArgumentKind.xml
- CustomAttributeTypedArgument`1.xml
- CustomAttributeValue`1.xml
- MemberReferenceKind.xml
- MethodSignature`1.xml
- PrimitiveSerializationTypeCode.xml
- PrimitiveTypeCode.xml
- SerializationTypeCode.xml
- SignatureAttributes.xml
- SignatureCallingConvention.xml
- SignatureHeader.xml
- SignatureKind.xml
- SignatureTypeCode.xml
- SignatureTypeKind.xml
- StandaloneSignatureKind.xml

The source code for these topics is located at:
https://github.com/dotnet/corefx/blob/master/src/System.Reflection.Metadata/src/System/Reflection/Metadata/Signatures

Issue #2487 

## Suggested Reviewers
@mairaw

### CODE REVIEW

File: xml/System.Reflection.Metadata/SignatureHeader.xml [hunk 1]
```diff
@@ -341,9 +348,9 @@
         <ReturnType>System.Byte</ReturnType>
       </ReturnValue>
       <Docs>
-        <summary>To be added.</summary>
-        <value>To be added.</value>
-        <remarks>To be added.</remarks>
+        <summary>Gets the raw value.</summary>
```
Comment (rpetrusha) [inline xml/System.Reflection.Metadata/SignatureHeader.xml]: "Gets the raw value of the header byte."

File: xml/System.Reflection.Metadata/SignatureHeader.xml [hunk 2]
```diff
@@ -341,9 +348,9 @@
...
+        <value>The raw value.</value>
```
Comment (rpetrusha) [inline xml/System.Reflection.Metadata/SignatureHeader.xml]: The raw value of the header byte.

File: xml/System.Reflection.Metadata/SignatureTypeCode.xml [hunk 3]
```diff
@@ -85,7 +85,7 @@
         <ReturnType>System.Reflection.Metadata.SignatureTypeCode</ReturnType>
       </ReturnValue>
       <Docs>
-        <summary>To be added.</summary>
+        <summary>Represents <see cref="System.Byte"/> in signatures.</summary>
```
Comment (rpetrusha) [inline xml/System.Reflection.Metadata/SignatureTypeCode.xml]: There should be an article for this and each of the following member descriptions -- e.g., "Repesents a Byte...."

File: xml/System.Reflection.Metadata/SignatureTypeCode.xml [hunk 4]
```diff
@@ -301,7 +301,7 @@
         <ReturnType>System.Reflection.Metadata.SignatureTypeCode</ReturnType>
       </ReturnValue>
       <Docs>
-        <summary>To be added.</summary>
+        <summary>Represents a <see cref="System.Object"/> in signatures.</summary>
```
Comment (rpetrusha) [inline xml/System.Reflection.Metadata/SignatureTypeCode.xml]: a -> an

**Review by rpetrusha — CHANGES_REQUESTED** [review]
@tompratt-AQ, a few comments. It would also be helpful if you could take a quick look at this, @tmat.

**Comment by tompratt-AQ** [comment]
@rpetrusha Thanks for the review Ron! I made all of your suggested changes.

**Review by rpetrusha — APPROVED** [review]

**Comment by tmat** [comment]
Not sure how this process works - wouldn't it be better to add the doc comments to the actual source code? What happens if we add them in future? Which comment wins?

**Comment by rpetrusha** [comment]
@tmat, doc comments in source code are a good starting point -- we can then use them to generate the initial documentation for a type/member. We don't have a mechanism for migrating documentation back to source comments, though, so once documentation is written, it exists independently of source code comments.
In this case, there were no comments in the source code for a number of the members and I didn't think that they were completely self-explanatory, so I was hoping that you could take a look at the docs to make sure the documentation is accurate.

**Comment by mairaw** [comment]
@tompratt-AQ can you also add a description to the namespace itself?

**Comment by mairaw** [comment]
@tmat @nguerrera if you guys want to review this, you can go to staging site at https://review.docs.microsoft.com/en-us/dotnet/api/system.reflection.metadata.constanttypecode?view=netcore-2.0&branch=pr-en-us-2630 and then navigate through the classes listed on the PR description

File: xml/System.Reflection.Metadata/ConstantTypeCode.xml [hunk 5]
```diff
@@ -175,7 +175,7 @@
         <ReturnType>System.Reflection.Metadata.ConstantTypeCode</ReturnType>
       </ReturnValue>
       <Docs>
-        <summary>To be added.</summary>
+        <summary>A null.</summary>
```
Comment (mairaw) [inline xml/System.Reflection.Metadata/ConstantTypeCode.xml]: wouldn't be better to say "A null reference?"

File: xml/System.Reflection.Metadata/MethodSignature`1.xml [hunk 6]
```diff
@@ -118,9 +118,16 @@
...
-        <summary>To be added.</summary>
-        <value>To be added.</value>
-        <remarks>To be added.</remarks>
+        <summary>Gets the number of parameters that are required for the method.</summary>
+        <value>The number of required parameters.</value>
+        <remarks>
+          <format type="text/markdown"><![CDATA[  
+  
+## Remarks  
+ This value will be equal to the length of the <xref:System.Reflection.Metadata.MethodSignature`1.ParameterTypes> property, unless this method signature represents the standalone call site of a vararg method, in which case, the extra entries in <xref:System.Reflection.Metadata.MethodSignature`1.ParameterTypes> are the types used for the optional parameters.
```
Comment (mairaw) [inline xml/System.Reflection.Metadata/MethodSignature`1.xml]: will be equal -> is equal

vararg should probably be inline code

File: xml/System.Reflection.Metadata/SignatureHeader.xml [hunk 7]
```diff
@@ -209,9 +216,9 @@
         <ReturnType>System.Boolean</ReturnType>
       </ReturnValue>
       <Docs>
-        <summary>To be added.</summary>
-        <value>To be added.</value>
-        <remarks>To be added.</remarks>
+        <summary>Indicates whether this <see cref="T:System.Reflection.Metadata.SignatureHeader"/> structure has the <see cref="T:System.Reflection.Metadata.SignatureAttributes.ExplicitThis"/> signature attribute.</summary>
```
Comment (mairaw) [inline xml/System.Reflection.Metadata/SignatureHeader.xml]: Boolean properties summaries start with Gets a value that indicates whether

File: xml/System.Reflection.Metadata/SignatureHeader.xml [hunk 8]
```diff
@@ -319,11 +326,11 @@
...
       </Parameters>
       <Docs>
-        <param name="left">To be added.</param>
-        <param name="right">To be added.</param>
-        <summary>To be added.</summary>
-        <returns>To be added.</returns>
-        <remarks>To be added.</remarks>
+        <param name="left">The first value to compare.</param>
+        <param name="right">The second value to compare.</param>
+        <summary>Determines whether two <see cref="T:System.Reflection.Metadata.SignatureHeader"/> values are unequal</summary>
```
Comment (mairaw) [inline xml/System.Reflection.Metadata/SignatureHeader.xml]: missing period

File: xml/System.Reflection.Metadata/SignatureTypeKind.xml [hunk 9]
```diff
@@ -46,7 +46,7 @@
         <ReturnType>System.Reflection.Metadata.SignatureTypeKind</ReturnType>
       </ReturnValue>
       <Docs>
-        <summary>To be added.</summary>
+        <summary>It is not known in the current context if the type reference or definition is a class or value type.</summary>
```
Comment (mairaw) [inline xml/System.Reflection.Metadata/SignatureTypeKind.xml]: It is not -> It isn't

**Review by mairaw — DISMISSED** [review]
LGTM. Left a few minor comments plus the request to add namespace description.

File: xml/System.Reflection.Metadata/ArrayShape.xml [hunk 10]
```diff
@@ -32,11 +32,11 @@
         <Parameter Name="lowerBounds" Type="System.Collections.Immutable.ImmutableArray&lt;System.Int32&gt;" />
       </Parameters>
       <Docs>
-        <param name="rank">To be added.</param>
-        <param name="sizes">To be added.</param>
-        <param name="lowerBounds">To be added.</param>
-        <summary>To be added.</summary>
-        <remarks>To be added.</remarks>
+        <param name="rank">The number of dimensions in the array.</param>
+        <param name="sizes">The sizes of each dimension.</param>
```
Comment (tmat) [inline xml/System.Reflection.Metadata/ArrayShape.xml]: >s [](start = 36, length = 1)

Singular?

File: xml/System.Reflection.Metadata/ArrayShape.xml [hunk 11]
```diff
@@ -32,11 +32,11 @@
...
+        <param name="lowerBounds">The lower-bounds of each dimension.</param>
```
Comment (tmat) [inline xml/System.Reflection.Metadata/ArrayShape.xml]: >s [](start = 49, length = 1)

Singular?

File: xml/System.Reflection.Metadata/ArrayShape.xml [hunk 12]
```diff
@@ -53,9 +53,9 @@
         <ReturnType>System.Collections.Immutable.ImmutableArray&lt;System.Int32&gt;</ReturnType>
       </ReturnValue>
       <Docs>
-        <summary>To be added.</summary>
-        <value>To be added.</value>
-        <remarks>To be added.</remarks>
+        <summary>Gets the lower-bounds of each dimension. Length may be smaller than rank, in which case the trailing dimensions have unspecified lower bounds.</summary>
```
Comment (tmat) [inline xml/System.Reflection.Metadata/ArrayShape.xml]: >each dimension [](start = 42, length = 14)

"all dimensions"?

File: xml/System.Reflection.Metadata/ArrayShape.xml [hunk 13]
```diff
@@ -53,9 +53,9 @@
...
+        <value>A collection of lower-bounds values.</value>
```
Comment (tmat) [inline xml/System.Reflection.Metadata/ArrayShape.xml]: >values [](start = 44, length = 6)

omit "values"

Comment (tmat) [inline xml/System.Reflection.Metadata/ArrayShape.xml]: >A collection [](start = 15, length = 12)

"An array"

File: xml/System.Reflection.Metadata/ArrayShape.xml [hunk 14]
```diff
@@ -91,9 +91,16 @@
         <ReturnType>System.Collections.Immutable.ImmutableArray&lt;System.Int32&gt;</ReturnType>
       </ReturnValue>
       <Docs>
-        <summary>To be added.</summary>
-        <value>To be added.</value>
-        <remarks>To be added.</remarks>
+        <summary>Gets the sizes of each dimension.</summary>
```
Comment (tmat) [inline xml/System.Reflection.Metadata/ArrayShape.xml]: >each dimension [](start = 35, length = 14)

"all dimensions"

File: xml/System.Reflection.Metadata/ArrayShape.xml [hunk 15]
```diff
@@ -91,9 +91,16 @@
...
+        <value>A collection of size values.</value>
```
Comment (tmat) [inline xml/System.Reflection.Metadata/ArrayShape.xml]: >A collection of size values [](start = 15, length = 27)

An array of sizes

File: xml/System.Reflection.Metadata/ArrayShape.xml [hunk 16]
```diff
@@ -91,9 +91,16 @@
...
+        <remarks>
+          <format type="text/markdown"><![CDATA[  
+  
+## Remarks  
+ The number of elements in the returned array may be fewer than the number of dimensions of the array type (its rank). In that case, the trailing dimensions have unspecified sizes.
```
Comment (tmat) [inline xml/System.Reflection.Metadata/ArrayShape.xml]: >fewer [](start = 53, length = 5)

smaller

Comment (tmat) [inline xml/System.Reflection.Metadata/ArrayShape.xml]: >s [](start = 178, length = 1)

singular

File: xml/System.Reflection.Metadata/ConstantTypeCode.xml [hunk 17]
```diff
@@ -12,8 +12,8 @@
     <BaseTypeName>System.Enum</BaseTypeName>
   </Base>
   <Docs>
-    <summary>To be added.</summary>
-    <remarks>To be added.</remarks>
+    <summary>Specifies constants that define types in a metadata type signature.</summary>
```
Comment (tmat) [inline xml/System.Reflection.Metadata/ConstantTypeCode.xml]: >Specifies constants that define types in a metadata type signature [](start = 13, length = 66)

Rather "Values representing types of metadata constants"

File: xml/System.Reflection.Metadata/ConstantTypeCode.xml [hunk 18]
```diff
@@ -31,7 +31,7 @@
         <ReturnType>System.Reflection.Metadata.ConstantTypeCode</ReturnType>
       </ReturnValue>
       <Docs>
-        <summary>To be added.</summary>
+        <summary>A <see cref="System.Boolean"/> type.</summary>
```
Comment (tmat) [inline xml/System.Reflection.Metadata/ConstantTypeCode.xml]: ><see cref="System.Boolean"/> [](start = 19, length = 28)

Not sure why we link System type here and in "Char" but not in other places.

File: xml/System.Reflection.Metadata/CustomAttributeNamedArgumentKind.xml [hunk 19]
```diff
@@ -12,8 +12,8 @@
     <BaseTypeName>System.Enum</BaseTypeName>
   </Base>
   <Docs>
-    <summary>To be added.</summary>
-    <remarks>To be added.</remarks>
+    <summary>Specifies constants that define the kinds of arguments in a custom metadata attribute.</summary>
```
Comment (tmat) [inline xml/System.Reflection.Metadata/CustomAttributeNamedArgumentKind.xml]: >custom metadata attribute [](start = 73, length = 25)

... in a custom attribute signature

File: xml/System.Reflection.Metadata/CustomAttributeNamedArgument`1.xml [hunk 20]
```diff
@@ -15,9 +15,9 @@
   </Base>
   <Interfaces />
   <Docs>
-    <typeparam name="TType">To be added.</typeparam>
-    <summary>To be added.</summary>
-    <remarks>To be added.</remarks>
+    <typeparam name="TType">The type of the argument.</typeparam>
+    <summary>Represents a named argument of the specified generic type for a custom metadata attribute.</summary>
```
Comment (tmat) [inline xml/System.Reflection.Metadata/CustomAttributeNamedArgument`1.xml]: >generic [](start = 58, length = 7)

omit "generic"

Comment (tmat) [inline xml/System.Reflection.Metadata/CustomAttributeNamedArgument`1.xml]: >Represents a named argument of the specified generic type for a custom metadata attribute [](start = 13, length = 89)

Represents a named argument decoded from a custom attribute signature

File: xml/System.Reflection.Metadata/CustomAttributeNamedArgument`1.xml [hunk 21]
```diff
@@ -15,9 +15,9 @@
   </Base>
   <Interfaces />
   <Docs>
-    <typeparam name="TType">To be added.</typeparam>
-    <summary>To be added.</summary>
-    <remarks>To be added.</remarks>
+    <typeparam name="TType">The type of the argument.</typeparam>
```
Comment (tmat) [inline xml/System.Reflection.Metadata/CustomAttributeNamedArgument`1.xml]: >The type of the argument [](start = 28, length = 24)

Type used to represent types of values decoded from the custom attribute signature.

File: xml/System.Reflection.Metadata/CustomAttributeNamedArgument`1.xml [hunk 22]
```diff
@@ -37,12 +37,12 @@
...
       </Parameters>
       <Docs>
-        <param name="name">To be added.</param>
-        <param name="kind">To be added.</param>
-        <param name="type">To be added.</param>
-        <param name="value">To be added.</param>
-        <summary>To be added.</summary>
-        <remarks>To be added.</remarks>
+        <param name="name">The name of the argument.</param>
+        <param name="kind">The kind of argument.</param>
```
Comment (tmat) [inline xml/System.Reflection.Metadata/CustomAttributeNamedArgument`1.xml]: the

File: xml/System.Reflection.Metadata/CustomAttributeNamedArgument`1.xml [hunk 23]
```diff
@@ -37,12 +37,12 @@
...
+        <param name="type">The type of argument.</param>
```
Comment (tmat) [inline xml/System.Reflection.Metadata/CustomAttributeNamedArgument`1.xml]: the

**Comment by tompratt-AQ** [comment]
@tmat 
@mairaw 
I made all of the suggested changes from both of you. There are just two remaining items:

1. In the[ xml/System.Reflection.Metadata/ConstantTypeCode.xml](https://github.com/dotnet/docs/pull/2630/files/694ed6a0e0139e75b062d0adb64d793370c3af7b#diff-7bfab040906f7ecf7b349f9710fa4486 ) file, I borrowed the summary descriptions for each type from the [CorElementType Enumeration1](https://docs.microsoft.com/en-us/dotnet/framework/unmanaged-api/metadata/corelementtype-enumeration). I did this because the code file for [ConstantTypeCode](https://raw.githubusercontent.com/dotnet/corefx/master/src/System.Reflection.Metadata/src/System/Reflection/Metadata/Signatures/ConstantTypeCode.cs) indicates that these values map to CorElementType values. Should I keep these descriptions, or should I use link to the CorElementType values?

2. Please review my namespace description for ns-System.Reflection.Metadata. 

Thanks to both of you for the detailed review!

**Pull request merged**